Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp

```diff
@@ -262,43 +246,39 @@ void ThreadedCompositor::renderLayerTree()
 
     // Retrieve the scene attributes in a thread-safe manner.
     WebCore::IntSize viewportSize;
-    float scaleFactor;
-    bool needsResize;
+    float deviceScaleFactor;
     uint32_t compositionRequestID;
     {
         Locker locker { m_attributes.lock };
         viewportSize = m_attributes.viewportSize;
-        scaleFactor = m_attributes.scaleFactor;
-        needsResize = m_attributes.needsResize;
+        deviceScaleFactor = m_attributes.deviceScaleFactor;
         compositionRequestID = m_attributes.compositionRequestID;
 
 #if !HAVE(DISPLAY_LINK)
         // Client has to be notified upon finishing this scene update.
         m_attributes.clientRendersNextFrame = m_scene->state().layersDidChange();
 #endif
-
-        // Reset the needsResize attribute to false.
-        m_attributes.needsResize = false;
     }
 
+    if (viewportSize.isEmpty())
+        return;
+
     TransformationMatrix viewportTransform;
-    viewportTransform.scale(scaleFactor);
+    viewportTransform.scale(deviceScaleFactor);
 
-    // Resize the client, if necessary, before the will-render-frame call is dispatched.
+    // Resize the surface, if necessary, before the will-render-frame call is dispatched.
     // GL viewport is updated separately, if necessary. This establishes sequencing where
     // everything inside the will-render and did-render scope is done for a constant-sized scene,
     // and similarly all GL operations are done inside that specific scope.
-
-    if (needsResize)
-        m_surface->clientResize(viewportSize);
+    bool needsGLViewportResize = m_surface->resize(viewportSize);
```

Comment:
If `deviceScaleFactor` is different than 1 shouldn't that require `needsGLViewportResize` to be true?

Reply:
viewportSize is already scaled, so it will be different when deviceScaleFactor changes